tests/e2e-tests/config/teardown.js

```diff
@@ -0,0 +1,20 @@
+/**
+ * Internal dependencies
+ */
+const {
+	deleteTaxes,
+	deleteCoupons,
+	deleteProducts,
+	deleteShippingZones,
+} = require( './fixture-loaders' );
+const { teardown: teardownPuppeteer } = require( 'jest-environment-puppeteer' );
+module.exports = async ( globalConfig ) => {
+	await teardownPuppeteer( globalConfig );
+	const { taxes, coupons, products, shippingMethods } = global.wc;
```

Comment:
Might be worth a comment or two to make it clear that [node `global`](https://nodejs.org/api/all.html#all_global) is used to communicate data between setup and teardown.

Or perhaps it's possible to do this some other way - e.g. use a single module for setup/teardown with module-local state.

Reply:
> Or perhaps it's possible to do this some other way - e.g. use a single module for setup/teardown with the module-local state.

I'm not sure I fully understand this, setup and teardown must be two functions, global is the only way to share scope between them, any code I declare at the top of the file (outside the function) will be ignored.

Reply:
The main feedback here is that we should avoid using global if possible. Not a big deal, probably not a blocker.

By "module" I mean js source file. In node you can have [module state](https://nodejs.org/api/all.html#globals_global) - local variables in a module can be accessed by all functions in the file, but not outside the module.

Right now the setup and teardown functions are each contained within their own modules, I'm not sure if it's practical to move them to a shared module. I see the jest config references them by filename rather than function object. In jest docs it looks like maybe they could maybe be defined in the same file - they are [passed as callbacks to APIs like `beforeAll`](https://jestjs.io/docs/en/setup-teardown#one-time-setup).

There are other ways to share state in JS without using global/window - e.g. using a closure.  For example, define the setup and teardown and their state in a closure:

```js
export function fixtureDataSetup() {
   var fixtureDataIds = {};

   function setup() {
     // make products etc & save ids
    fixtureDataIds.products = ...;
  }
  
  function teardown() {
     // delete stuff
     removeProducts( fixtureDataIds.products );
  } 

  return { 
     setup,
     teardown,
  };
}

// And then out in the main / config you'd tell Jest to run `setup` and `teardown` as global setup/teardown func.
```

This might be overkill - just exploring options. I could imagine a world where the test data ids would be useful to tests - e.g. a test that submits a review for a particular product id. In that case, the closure variable could be exposed (directly, or via getter funcs etc).

If we stick with the global I think it's a good idea to add a big comment explaining your intentions for it – global variables need big comments :)

Reply:
jest expects setup and teardown to be the default export on that file, there's a lot of care of why they shouldn't context, I can see your example and I think it might work but I do believe its a bit overkill for this exact scenario (and would make code readability a bit jumpy between three files).

In the context of Jest, teardown and setup run on a separate context compared to actual test and code, there is no sharing, I can't access the same global nor can I access the same `.wc` variable, so in that small context, I think it's really safe.
Using global is the recommended way to handle shared data in Jest documentation when it comes to those two functions.

Reply:
>  test that submits a review for a particular product id.

This is not really possible with how jest work since this data is not accessible to it, I was thinking of ways we can circulate around this, in the rare cases we need such a thing, we can initiate our fixture from within the test, so, inside a `beforeAll`, this will require us to change the way our loaders work (by receiving fixtures as a prop).